tested/dsl/ast_translator.py

```diff
@@ -333,6 +335,24 @@ def _translate_to_ast(node: ast.Interactive, is_return: bool) -> Statement:
         return _convert_statement(statement_or_expression)
 
 
+def extract_comment(code: str) -> str:
+    """
+    Extract the comment from the code.
+
+    :param code: The code to extract the comment from.
+    :return: The comment if it exists, otherwise an empty string.
+    """
+    tokens = tokenize.tokenize(BytesIO(code.encode("utf-8")).readline)
```

Comment:
The Python module also has a function `tokenize.generate_tokens` that takes a string instead of a bytes array, which avoids the extra encoding step:

```
tokens = tokenize.generate_tokens(io.StringIO(code).readline)
```

Reply:
If we assume that a TESTed statement or expression does not contain multiple expressions or statements (check is needed to see if this is supported by TESTed and if TESTed iterates over these expressions/statements itself instead of relying on language modules to convert sequential blocks of statements) we may only process comments at the end of the statement/expression. If a single expression/statement is split over multiple lines, as is possible in Python, it may be possible to include comments at the end of every line. However, it is not guaranteed that this is supported by other languages. Instead of preventing the use of inline comments in the generic language of TESTed, we might simply ignore them when making language-specific conversions (in the language modules) and only process the end-of-statement/expression comment if present.

It seems the Python tokenize always has a `tokenize.NEWLINE` and a `tokenize.ENDMARKER` at the end of the token stream, so the above comes down to checking if the third token from the end (if stream has at least three tokens) is a comment-token:

```python
comment = ''
tokens = tuple(tokenize.generate_tokens(io.StringIO(code).readline))  # TODO: scan token stream without tuple conversion
if len(tokens) and tokens[-3].type == tokenize.comment:
    comment = tokens[-3].string.lstrip('#').rstrip()
return comment
```

Reply:
I checked if tokenization and parsing could share some of their processing steps, but this does not seem to be the case. We should at least measure the impact of the extra tokenization step for each expression/statement in the DSL on the overall performance of TESTed. Performance of generating feedback is an important feature of TESTed.

Reply:
> If we assume that a TESTed statement or expression does not contain multiple expressions or statements (check is needed to see if this is supported by TESTed and if TESTed iterates over these expressions/statements itself instead of relying on language modules to convert sequential blocks of statements)

I just tried this in the following way:
```yaml
- tab: sort_words
  contexts:
    - testcases:
      - statement: 'result: map = count_words(["Gobber", "Hiccup", "Astrid", "Stoick", "Toothless"], "httyd.txt") 
                    result2 = {"test": 10}'
      - expression: 'sort_words(result2) # This is an expression'
        return: ["Hiccup", "Toothless", "Stoick", "Astrid", "Gobber"]
        files:
          - name: "httyd.txt"
            url: "httyd.txt"

```
This generates an error while trying to validate the DSL. So this is as expected. The exact same thing also happens for expressions. When you put a comment in between the 2 lines, TESTed will think that the second line is part of the comment and everything will execute as normal.

I also tried the following:
```yaml
- tab: sort_words
  contexts:
    - testcases:
      - statement: 'result: map = count_words(["Gobber", # This is a name
                                               "Hiccup", # This is also name
                                               "Astrid", # Wow another name
                                               "Stoick", # You guessed it... A NAME
                                               "Toothless" # NAME
                                               ], "httyd.txt") # This is a statement'
      - expression: 'sort_words(result)'
        return: ["Hiccup", "Toothless", "Stoick", "Astrid", "Gobber"]
        files:
          - name: "httyd.txt"
            url: "httyd.txt"

```
While this works for python it does not seem to work when validating the DSL. It can't find the closing `]`, which suggests that everything from the first `#` is a comment. I think it might be for the best to just put these at the end of a statement.

Reply:
See https://yaml-multiline.info/ for info about multiline strings in YAML.

Reply:
Ow, that does work. All the comments are picked up by the tokenizer when doing it the following way:
```yaml
- tab: sort_words
  contexts:
    - testcases:
      - statement: >
               result: map = count_words(["Gobber", # This is a name
                                          "Hiccup", # This is another name
                                          "Astrid", # This yet another name
                                          "Stoick", # WOOOO!!! A name!
                                          "Toothless" # and the last name
                                          ], "httyd.txt") # This is a statement
      - expression: 'sort_words(result) # This is an expression'
        return: ["Hiccup", "Toothless", "Stoick", "Astrid", "Gobber"]
        files:
          - name: "httyd.txt"
            url: "httyd.txt"

```
The test also run without any problem. Comments could potentially be shown this way. The mayor challenge would be how to store the location of these comments in the new representation of the testcase.

Reply:
This problem is resolved for now. We could always come back to this at a later time.